sky/skylet/job_lib.py

```diff
@@ -540,57 +537,45 @@ def update_job_status(job_ids: List[int],
-
     job_client = _create_ray_job_submission_client()
 
-    # In ray 2.4.0, job_client.list_jobs returns a list of JobDetails,
-    # which contains the job status (str) and submission_id (str).
-    job_detail_lists: List['ray_pydantic.JobDetails'] = job_client.list_jobs()
-
-    pending_jobs = _get_pending_jobs()
-    job_details = {}
-    ray_job_ids_set = set(ray_job_ids)
-    for job_detail in job_detail_lists:
-        if job_detail.submission_id in ray_job_ids_set:
-            job_details[job_detail.submission_id] = job_detail
-    job_statuses: List[Optional[JobStatus]] = [None] * len(ray_job_ids)
-    for i, ray_job_id in enumerate(ray_job_ids):
-        job_id = job_ids[i]
-        if ray_job_id in job_details:
-            ray_status = job_details[ray_job_id].status
-            job_statuses[i] = _RAY_TO_JOB_STATUS_MAP[ray_status]
-        if job_id in pending_jobs:
-            if pending_jobs[job_id]['created_time'] < psutil.boot_time():
-                logger.info(
-                    f'Job {job_id} is stale, setting to FAILED: '
-                    f'created_time={pending_jobs[job_id]["created_time"]}, '
-                    f'boot_time={psutil.boot_time()}')
-                # The job is stale as it is created before the instance
-                # is booted, e.g. the instance is rebooted.
-                job_statuses[i] = JobStatus.FAILED
-            # Gives a 60 second grace period between job being submit from
-            # the pending table until appearing in ray jobs.
-            if (pending_jobs[job_id]['submit'] > 0 and
-                    pending_jobs[job_id]['submit'] <
-                    time.time() - _PENDING_SUBMIT_GRACE_PERIOD):
-                # For jobs submitted outside of the grace period, we will
-                # consider the ray job status.
-                continue
-            else:
-                # Reset the job status to PENDING even though it may not appear
-                # in the ray jobs, so that it will not be considered as stale.
-                job_statuses[i] = JobStatus.PENDING
-
-    assert len(job_statuses) == len(job_ids), (job_statuses, job_ids)
-
     statuses = []
-    for job_id, status in zip(job_ids, job_statuses):
+    for job_id in job_ids:
+        ray_job_id = make_ray_job_id(job_id)
         # Per-job status lock is required because between the job status
         # query and the job status update, the job status in the databse
         # can be modified by the generated ray program.
         with filelock.FileLock(_get_lock_path(job_id)):
+            status = None
+            try:
+                job_detail = job_client.get_job_info(ray_job_id)
+            except RuntimeError:
+                # An RuntimeError is raised when the job does not exist.
+                # https://docs.ray.io/en/latest/cluster/running-applications/job-submission/doc/ray.job_submission.JobSubmissionClient.get_job_info.html  # pylint: disable=line-too-long
+                pass
```

Comment:
How long will this take? It can be quite costly to query ray job status for each job separately. Do you think it is possible to call the `list_jobs` outside the lock and for loop once? It may sacrifice a bit of the freshness of the information we get for the job, but it might be ok if we change `time.time()` in L573 to a constant we retrieve once outside the for loop?

Reply:
Good point! Updated.